win_service recovery actions support

rnsc opened the pull request:
##### SUMMARY
This is a new feature for the win_service Ansible module.
I wanted to be able to set actions related to the "Recovery" part of the service management GUI.
What is set in the GUI translates to a REG_BINARY key in the registry.
My code builds a hexadecimal combination of all the value that you passed through Ansible.
I'm comparing the values if it was already set for idempotency, and check mode is supported.

These settings can also be set with the sc.exe program, but I wanted to have something that was idempotent.

As far as variables naming, etc... I'm open to making changes you deem necessary. I commented as much as I saw fit, but if you feel like it's lacking, please do tell.

I've seen two schools when it comes to passing stuff to Powershell modules, either separate variables or a string that represents an Ansible dictionary, I went with the former. (Although I wished we could just pass a dict to a powershell script, it'd be more readable)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_service.py

##### ADDITIONAL INFORMATION
I did this for a personal need, so I'm using it in my local Ansible library directory.
I wanted to contribute back to the community in case someone else wanted this.

rnsc committed:
win_service recovery actions support

rnsc commented:
I'll fix the validation errors.

rnsc committed:
fixing validation issues

rnsc committed:
ansible_metadata shouldn't be changed...

rnsc committed:
blank line contains whitespace

rnsc committed:
pslint compliance!

rnsc committed:
whitespace

rnsc commented:
So I think the one piece missing here is the integration test for this specific feature, right?
Any advice for creating those?

rnsc commented:
ready_for_review

rnsc commented:
Hey @chrishoffman @dagwieers 

Can you check this whenever you have time? :)
I'd appreciate feedback on this PR. Although it works for me for my project, I'd very much like to have this feature added for the benefit of the community.

Regards,
rnsc

nitzmahone reviewed (changes requested):
I like the idea of adding support for this to round out Ansible's Windows service management capability, but I do have concerns (some conceptual, some with the specific implementation here):

* At a glance, it doesn't appear that this implementation is idempotent, which would be absolutely required
* No tests
* From a UI perspective, there are enough args around this that we might want to consider putting them in a sub-dictionary (esp since the new PS/C# arg validation stuff would make it easier to validate that now)
* TMK, the registry format for this is undocumented, which means manual changes to it are dangerous... There are Win32 APIs/structs available for managing it (see https://docs.microsoft.com/en-us/windows/desktop/api/winsvc/ns-winsvc-service_failure_actionsw), which is probably the way it should go. That said, if the other idempotence/UI/test issues are resolved, we could go forward with the manual registry hacking approach temporarily/experimentally, then replace it with a Win32-backed approach later if needed.

@jborean93 any other thoughts?

rnsc commented:
@nitzmahone thank you for reviewing this PR!
And thanks for the feedback!
To answer your concerns, this is idempotent, I wanted something idempotent to avoid using the sc.exe through win_shell.
I do agree that it looks hackish, and it’s not very well documented, but from my testing, the way this reg entry is built is always the same.

Not sure I understand the sub dictionary part. I wanted to do that, but powershell doesn’t seem to be handling a dictionary passed by Ansible too well (maybe I didn’t do it properly), but with a bit more guidance I’d be happy to change that too.

As for the testing, I wasn’t sure I had found the appropriate playbook to update, so again, if you have a tip, I’d be happy to add tests to make it this PR better.

jhawkesworth commented:
I agree this would be nice to have, and that a separate sub-dict for recovery options would be nice from a readability point of view.

I have some suggestions for the names of the new parameters.  Its not allways possible but I like it if the names can be as similar as possible to the names you see in the GUI.  This makes it much easier for users to automate what they are seeing in the user interface.  A case in point: I'm not sure what 'reset_period' relates to in the Recovery UI.  Likewise I'm not keen on using the bare 'magic' numbers for the recovery actions (0.1.2.3) but would rather have a controlled list of allowed names, say something like: 
no_action, restart_service, run_command, restart_computer

Here's an example of what I think a subdict might look like:

```- name: Set recovery actions
  win_service:
    name: service name
    recovery:
      reset_fail_count_after: 300	
      on_first_failure: restart_service
      first_failure_timeout: 60000
      on_second_failure: restart
      second_failure_timeout: 60000
      on_subsequent_failure: restart_computer
      subsequent_failure_timeout: 60000	
``
Hope the above is helpful,

Jon

rnsc commented:
Hello @jhawkesworth 

Thank you for the feedback, it was indeed helpful!
Initially, I did try to go for that approach, because it was way more readable, but it didn't work in Powershell properly. I'll have another go at it. But I've seen other people complaining that using YAML dictionaries with Powershell modules in Ansible is tricky (and they resort to using a dictionary as a string in YAML to parse it as is in Powershell).

As far as naming goes, that's something I also agree on, I'll check what I can do to make it more relevant in regards to the naming in the Windows GUI properties dialog.
The "magic" numbers is something a colleague pointed out to me and I've already changed it in my internal version of the module. I just haven't had the time to update the present PR with that change.

I'll also try to add the tests that are lacking.

jborean93 reviewed (changes requested):
I'm with @nitzmahone here, the points he's laid out are exactly what I would like to see here. Manually changing the registry is worrisome, using PInvoke is far better but can get complex, if `sc.exe` exposes the options then that is far more desirable than editing the registry keys directly.

I could also see this as a separate module, `win_service_recovery`, but I'm not married to the idea. This is mostly because win_service is quite large and makes things nice and simple to test and maintain going forward. In saying that, if the work is done in `win_service` instead of a separate module I could be ok with that but it would have to address the points nitzmahone has brought up.

rnsc commented:
hello @jborean93 

I do understand your concern about setting reg key directly instead of sc.exe, BUT, to recover the settings the query system of sc.exe means parsing the output of the command line (and it's not pretty).
To me it looks as hacky as reg key, but I'll give it a try and I trust that we can agree on which one is best afterwards.
I could also have sc.exe for the set, and retrieve the reg key to compare for idempotency, but that would still involve a reg key and doing hex conversion, which I'm not sure you'll like.

![image](https://user-images.githubusercontent.com/1765108/58227007-5e419c00-7d29-11e9-9948-724a3bdb53a2.png)

jborean93 commented:
@rnsc that's why PInvoking was the alternative option, looks like sc is pretty poor to try and parse though :(.

rnsc committed:
Added support for Ansible dict + using sc.exe to set the values instead of reg

rnsc committed:
Updated the doc to reflect the changes

rnsc committed:
Return a proper dict to Ansible with the recovery settings

rnsc committed:
Updated the documentation with fixes and returned values

rnsc committed:
Fixing the documentation

rnsc committed:
Fixing whitespaces

rnsc commented:
Hello @jborean93 @nitzmahone @jhawkesworth 

I updated the PR with a few changes.
You can now pass a dict (I don't know how I tested the farm, because it worked out of the box).

I'm using SC.EXE to set the variable, but I'm still reading the key from the registry to compare the values for idempotence and returning them to Ansible.
Do you want me to go the extra-mile and try to parse the output of SC.EXE query mechanism?

I updated the documentation according to the latest changes.

I'm now looking at adding integration tests.

rnsc committed:
Adding integration test for win_service recovery settings

rnsc commented:
I just added integration tests.

rnsc committed:
Fixing task names and variables checks

rnsc committed:
Fixing mapping of actions

rnsc committed:
Updating the doc to reflect code changes

rnsc committed:
Removing hardcoded value and checking for empty recovery string

rnsc committed:
Fixing the again variable for idempotency

rnsc committed:
fixing the sample data in the documentation

imjoseangel commented:
From my side just saying the I totally agree with @jborean93. Pinvoking is the best option and I think a separate module too. Although it is quite challenging I vote for those two options.

rnsc commented:
hello @imjoseangel 

I've actually made all the changes that were requested. Excepted for separating this into its own module.
For me it makes sense to keep this in the win_service module as it's part of service properties and definitions.
It doesn't make the module more complex, it just adds functionalities. Although I do understand the fact that you might want to keep modules as small as possibles, I feel like splitting things too much might not be ideal from a UX pov. (my 2 cents)

@nitzmahone I looked at the Win32 API approach, and using Powershell to query it doesn't look trivial (at least for me).
@jborean93 Let me know if you like the new approach I took or if it's not good either.

jhawkesworth commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,47 @@
+      on_first_failure:
+        description:
+          - Action to take on first failure of the service.
+          - Only restart and reboot are supported value.
```
I suggest using the 'choices' and 'default' documentation elements here as they will format better in the module documentation page. 
for example, below 'description' add  'choices' and 'default' elements.  See win_xml.py and other windows modules for examples

```
        description:
          - Action to take on first failure of the service.
        choices: [ restart, reboot, no_action ]
        default: no_action
```

rnsc replied:
Thanks for the suggestion, I changed it!

jhawkesworth commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,47 @@
+        description:
+          - Action to take on second failure of the service.
+          - Only restart and reboot are supported value.
+          - Defaults to no_action.
```
As above

jhawkesworth commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,47 @@
+        description:
+          - Action to take on subsequent failure of the service.
+          - Only restart and reboot are supported value.
+          - Defaults to no_action.
```
As above

nitzmahone commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -371,6 +455,83 @@ Function Set-ServiceState($svc, $wmi_svc, $state) {
+
+    if ($recovery_actions -ne $processed_actions) {
+
+            $cmd_output=sc.exe \\$env:computername failure $name reset= $reset actions= $actions_string
```
This needs to be wrapped in a `$check_mode` conditional, otherwise it's potentially persisting changes in check mode...

nitzmahone replied:
Should be able to do this without the computername arg- the command likely no-ops it anyway, but if for some reason it *doesn't*, it will likely fail if it goes into a path that wants remote credentials. Probably better to just skip that arg altogether...

nitzmahone commented on test/integration/targets/win_service/tasks/tests.yml:
```diff
@@ -910,3 +910,53 @@
   failed_when: "fail_pause_stopped_service.msg != 'failed to pause service ' + test_win_service_name + ': The service does not support pausing'"
+
+- name: set recovery settings for the service
+  win_service:
```
Should probably add a `check_mode: yes` version of this test first, too, just to ensure that we're not persisting any of these changes (copying/anchoring this test and its assert is fine, as a failure that writes changes should cause the non-check-mode version not to change and thus fail)

rnsc replied:
Add a check block

nitzmahone reviewed (changes requested):
A few small things to fix... The way you're having `sc.exe` set the value makes me a lot more comfortable with this, since if the format ever *does* change, the worst thing is that the module is no longer idempotent for this part of the args, rather than corrupting the SCM DB.

If someone doesn't specify all the values, it will bomb right now, but that's probably OK- whenever we cut this over to the new .NET argspec impl, we'll get that suboption validation for free. The underlying SCM API also supports a lot more granularity (eg arbitrary action counts), but if we want to support that later, it's easy enough to make the input option type `raw` so we can sniff for a list or dict. So I'm OK with having the module UI match the Service snap-in UI for now...

nitzmahone commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,47 @@
+        description:
+          - Reset failure counter after this number of seconds.
+        type: int
+      on_first_failure:
```
Was talking this over with @jborean93 after I went looking at the underlying Win32 API for this, and the way `sc.exe` sets it as well, and we were thinking it might be better if, rather than mimic'ing the SCM UI for this, if it took a list of actions. That allows for shorter descriptions, like:
```
recovery_reset_interval: 300
recovery_actions:
- action: restart
  delay: 60000
```

but it means the registry side of things needs to be a little more dynamic, since cActions would vary with the size of the list. It also makes the return value a little simpler, since it'd just be a list. Thoughts?

Sephtex replied:
The recovery_actions been passed as a list seems much nicer indeed, and not forget shorter. Although when you think about the fact that a list is in theory not ordered (even if I've seen that on many occasions the order doesn't change), would it be a good idea to do it? Are there assurances that can be given in order to be sure a list stays in order?

jborean93 replied:
A list should be ordered, it's a dict that doesn't have a guaranteed order to the keys.

rnsc replied:
I do agree that a dynamic list would be easier to document and use.
I'll check what I can do to make it more dynamic. It also makes sense.

It would also make it less "bomby" in case of parameter not passed.

But changing this, would require me to redo a bunch of tests for idempotency, so if it happens, it'll take me some time.

rnsc replied:
@nitzmahone @jborean93 so changing it to a list is gonna take more time than I thought, not sure if I'll be able to make for the code freeze of Ansible 2.9.
Do you think this can already be a working solution for 2.9 and we can iterate on it for the next releases?

jborean93 replied:
@rnsc sorry for the delay, I would prefer not to implement a change then change the UI for those options in the future. The freeze isn't until the end of August, if you think you won't be able to make it let me know and I'll try to pick up where you've left off and see if we can get it in before that timeframe.

rnsc replied:
Hello @nitzmahone and @jborean93 
I implemented the changes you wanted to see, let me know if it's better now :)

rnsc committed:
Updating the documentation with choices

rnsc committed:
Adding a check_mode wrapper

rnsc committed:
Adding check mode test

rnsc committed:
Forgot to put last exit code check in check_mode wrapper

rnsc committed:
No values are set in check mode!

movergan commented:
Hi, Is there any progress? It would be really nice to have this feature!

beatcracker commented:
Can we also have [sc failureflag](https://serverfault.com/a/634180/270504) functionality? Under the hood it sets registry key `FailureActionsOnNonCrashFailures` to the specified `failureflag` value.

rnsc commented:
> Hi, Is there any progress? It would be really nice to have this feature!

Unfortunately, I haven't had time to work on this one recently.

rnsc committed:
comply with nitzmahone and jborean comments

rnsc commented:
I've updated the code and will let Shippable test build my changes :)

rnsc committed:
Tabs to spaces...

rnsc committed:
Updated example

rnsc committed:
Updated recovery actions tests

rnsc committed:
Version added should be 2.10

rnsc commented:
ready_for_review

rnsc committed:
force users to specify the three recovery states

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,20 @@
+    description:
+      - Reset failure counter after this number of seconds.
+    type: int
+    default: 1
```
Why not make it the default Windows uses?

rnsc replied:
I'll double check what the default is and change this.

rnsc replied:
There's no default in Windows, actually the GUI shows "0".
And you have to set it when setting actions.

ShachafGoldstein replied:
It seems the value is in days at least in win10 / server 16 and the defautl is 1 day.
in hex it is saved in milliseconds so 1 day = 80510100

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -409,6 +576,10 @@ Function Set-ServiceConfiguration($svc) {
     }
+
+    if (($null -ne $recovery_actions) -and ($null -ne $recovery_reset_interval)) {
+        Set-ServiceRecovery -name $svc.Name -reset_interval $recovery_reset_interval -actions $recovery_actions
```
Why not allow setting either one?

rnsc replied:
Well the thing is, if you set just one with sc.exe, it sets the same actions for the two other actions too.
If you set two, the last actions has the same value as the second one.
Unless I tested improperly. The behaviour is different when you set it in the UI.

We could allow for this to work if we used the registry to set the values, but @jborean93 was strongly disagreeing with that.

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -371,6 +456,88 @@ Function Set-ServiceState($svc, $wmi_svc, $state) {
+
+    $reset_fail_count_after = ReversedHexValue $reset_interval
+
+    if ($actions.Count -ge 3) {
```
Why allow more than 3?

rnsc replied:
That's an error on my part!

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,20 @@
+    description:
+      - A list representing the recovery actions you want to set.
+      - First element of the list is the first failure actions/delay...
+      - This can go up to a count of three for subsequent failures.
```
In the code, you enforce setting all 3

rnsc replied:
Indeed, I'll change the wording.

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -371,6 +456,88 @@ Function Set-ServiceState($svc, $wmi_svc, $state) {
+
+    if ($recovery_actions -ne $processed_actions) {
+
+            if (!$check_mode) {
```
Please use `-not` for readability

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -371,6 +456,88 @@ Function Set-ServiceState($svc, $wmi_svc, $state) {
+    if ($recovery_actions -ne $processed_actions) {
+
+            if (!$check_mode) {
+                $cmd_output=sc.exe failure $name reset= $reset actions= $actions_string
```
Why getting from the registry but setting with sc?

rnsc replied:
As discussed with @jborean93 , parsing the output of sc.exe isn't ideal either...
So I ended up with a hybrid solution that retrieves the values from the registry (which is easier imho, albeit you need some logic to read it).

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -144,6 +214,21 @@ Function Get-DelayedStatus($name) {
+      $recovery_actions = ""
+      $bytes = (Get-ItemProperty -LiteralPath $recovery_actions_key).FailureActions
+      $bytes | ForEach-Object { $recovery_actions+=('{0:x2}' -f $_).ToString() }
+      $recovery_actions = $recovery_actions.substring(0,$recovery_actions.length)
```
Why is this needed?

rnsc replied:
I don't remember, I'll test this again and add a comment in the code

rnsc replied:
It was not needed :)

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -144,6 +214,21 @@ Function Get-DelayedStatus($name) {
+      $recovery_actions = $false
+  }
+
+  $recovery_actions
```
Please use return otherwise it might cause issues if piped since it outs as output

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -88,6 +149,14 @@ Function Get-ServiceInfo($name) {
+    } else {
+        $recovery = @{}
+    }
+
```
Why not simply return `$null` or @{} in teh function?

rnsc replied:
Because I use that function elsewhere and I don't want the conversion to "Human"

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -59,6 +61,65 @@ if ($null -ne $path) {
+    $res
+}
+
+Function RecoveryActionsToHuman ($s) {
```
Please change it to be in Powershell format `convert-RecoveryActionsToHuman` maybe

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -59,6 +61,65 @@ if ($null -ne $path) {
+                            'subsequent_failure_timeout' = $res[10];
+                        }
+
+    $recovery_actions
```
Please use return

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -59,6 +61,65 @@ if ($null -ne $path) {
+        default { 0 }
+    }
+
+    $res
```
Please use return

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -59,6 +61,65 @@ if ($null -ne $path) {
+  $res
+}
+
+Function RecoveryActionMapping ($a) {
```
Why not using a hashtable?

rnsc replied:
And then I do lookups in that hashtable with the value I want to map?

ShachafGoldstein replied:
yes

rnsc replied:
I tested the hashtable option and couldn't make it work.

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -59,6 +61,65 @@ if ($null -ne $path) {
+
+  $res = [convert]::toint32($res, 16)
+
+  $res
```
Please use return

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -59,6 +61,65 @@ if ($null -ne $path) {
+
+  $res = [system.string]::Join('', $res)
+
+  $res
```
Please use return

ShachafGoldstein commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -23,6 +23,8 @@ $path = Get-AnsibleParam -obj $params -name 'path'
 $start_mode = Get-AnsibleParam -obj $params -name 'start_mode' -type 'str' -validateset 'auto','manual','disabled','delayed'
 $state = Get-AnsibleParam -obj $params -name 'state' -type 'str' -validateset 'started','stopped','restarted','absent','paused'
 $username = Get-AnsibleParam -obj $params -name 'username' -type 'str'
+$recovery_reset_interval = Get-AnsibleParam -obj $params -name 'recovery_reset_interval' -type 'int' -default 1
```
Consider changing the default to windows's default

rnsc committed:
updated documentation

rnsc committed:
fixing the code after review

rnsc commented:
Hello @ShachafGoldstein ,

I answered your questions and updated the code in accordance to your remarks.

Kind regards,
rnsc

rnsc commented:
ready_for_review

rnsc committed:
Substitute RecoveryActionMapping function for a hashtable

rnsc committed:
Missing parenthesis

rnsc committed:
Rolling back to function because hashtable doesn't work

rnsc commented:
ready_for_review

jborean93 commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,20 @@
+    version_added: '2.10'
+  recovery_actions:
+    description:
+      - A list representing the recovery actions you want to set (3).
```
I would omit the 3, it seems like in the internal API you can set as many recovery actions as you want. It's just the GUI that limts to 3.

jborean93 commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,20 @@
+  recovery_actions:
+    description:
+      - A list representing the recovery actions you want to set (3).
+      - First element of the list is the first failure actions/delay...
```
I would simplify this by saying the list order is the order of actions to take, i.e. first entry is the first action and so on.

jborean93 commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,20 @@
+    description:
+      - A list representing the recovery actions you want to set (3).
+      - First element of the list is the first failure actions/delay...
+      - Third element is for the subsequent failures.
```
Not needed.

jborean93 commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,20 @@
+      - A list representing the recovery actions you want to set (3).
+      - First element of the list is the first failure actions/delay...
+      - Third element is for the subsequent failures.
+    type: list
```
This takes a list but each entry is a dict. We should define the options in each dict. We can do this by adding

```
type: list
suboptions:
  action:
    description:
    - The description of the action
    choices:
    - restart
    - reboot
    - run
    required: yes
    type: str
  delay:
    description:
    - The description of delay
    type: int
    required: yes
```

The values of `suboptions` is just like the normal module options at the root level, so you can set things like `description`, `type`, `choices`, `default`, `required`, `version_added` and so forth. Doing this allows people to see what they can specify for each action and what it actually means rather than trying to decipher the examples alone.

jborean93 commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -113,6 +113,20 @@
+    default: 0
+    version_added: '2.10'
+  recovery_actions:
+    description:
```
We should add a note by saying win_service will run idempotency checks on the actions as a whole and not as individual entries. This is because the list order is important to determine the order of the actions themselves. Hopefully this makes sense?

Also what happens when someone specifies an empty list. In my mind this should just remove the recovery actions from the service but the code seems like it expects 3 or fails when not null.

jborean93 commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -292,4 +318,18 @@
     returned: success and service exists
     type: list
     sample: false
+recovery:
```
The recovery actions should also reflect the input way, i.e. returns a list of actions rather than a dict with `<x>_failure_<y>` and so forth. Basically it would just look similar to how you are defining `recovery_actions` as a module option right now.

jborean93 commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -59,6 +61,65 @@ if ($null -ne $path) {
+        $res[$i]=DecValue($res[$i])
+    }
+
+    $recovery_actions = @{
```
We should be returning a list of recovery actions not just 1 dict. If there are no recovery actions then returning an empty list is probably ideal.

jborean93 commented on lib/ansible/modules/windows/win_service.py:
```diff
@@ -229,6 +243,18 @@
+    recovery_reset_interval: 300
+    recovery_actions:
+      - action: restart
+        delay: 60000
```
When documenting these sub options make sure you specify the units, it looks like this is in milliseconds but I'm not 100% sure.

jborean93 commented on lib/ansible/modules/windows/win_service.ps1:
```diff
@@ -371,6 +456,88 @@ Function Set-ServiceState($svc, $wmi_svc, $state) {
+
+    $reset_fail_count_after = ReversedHexValue $reset_interval
+
+    if ($actions.Count -eq 3) {
```
Why have a hard count here, we should just set 1-3 actions based on the input ~fill the remaining options with Take No Action~ if it's less than 3. Maybe fail if it's more than 3. Also if it's an empty list we should remove all actions ~probably set Take No Action for all 3.~

Edit: see https://github.com/ansible/ansible/pull/56445#discussion_r346600024 for more info

Sephtex replied:
In https://github.com/ansible/ansible/pull/56445#discussion_r345020105 you mentioned that 3 is just a GUI limitation. And if the third element is for the subsequent failures, does that mean that the last option in the list (3rd in this case), the last action set, is meant for the subsequent actions for failures?
But it's still confusing though as your top comment would say to fill the actions up as 'Take no action', meaning, if you just set 1 (or even till 2), no action would be taken afterwards. But if you go to the 3rd action, it'll repeat for all the next. Does that mean if you set the action on entry 6 (in theory), the 6th time would count for 6 or more subsequent failures to take that same action?

jborean93 replied:
> does that mean that the last option in the list (3rd in this case), the last action set, is meant for the subsequent actions for failures?

That is correct, if you read the [Remarks](https://docs.microsoft.com/en-us/windows/win32/api/winsvc/ns-winsvc-service_failure_actionsw) for the failure actions structure you can see the following line

> If N is greater than cActions, the service controller repeats the last action in the array.

So say you have 2 actions define, 1 to run a command and the 2nd to restart the service it will do 1 then 2 and then subsequently repeat 2 forever.

> But it's still confusing though as your top comment would say to fill the actions up as 'Take no action

Sorry that's my misunderstanding there and I was wrong when I said that. Here is the various scenarios I see with `recovery_actions`:

* `recovery_actions: null` (or just not set): Do not do anything, keep the existing actions as is
* `recovery_actions: []`: Make sure we remove all recovery actions if any exist
* `recovery_actions: [action1, action2]`: Make sure we have 2 actions set, that make action1 and action2
* `recovery_actions: [action1, action2, action3]`: Same as above but we have 3 actions now
* `recovery_actions: [1, 2, 3, 4]`: For now we need to fail as `sc.exe` doesn't support this but in the future when we move to a PInvoke impl we can remove this limit

I've tested it out and `sc.exe failure` can be used to set 1, 2, or 3 actions not just 0 or 3. So the code just needs to make sure we check the existing settings properly and then set it if there is a change.

> Does that mean if you set the action on entry 6 (in theory), the 6th time would count for 6 or more subsequent failures to take that same action?

Yes that is correct, althrough `sc.exe failure` only allows you to set 3 you *can* set more by using the Win32 API which is a lot more involved.

jborean93 replied:
Just an FYI, I was playing around with it a bit more yesterday and created a barebones cmdlet to overcome the 3 action limit https://gist.github.com/jborean93/889288b56087a2c5def7fa49b6a8a0ad. Through this I was able to verify that I can set 4 or more actions and Windows will continue to work through them and finally repeat the last action.

rnsc replied:
At this point, we might as well just drop this PR and create a separate module that implements the Win32 API for Windows services recovery actions then. 😃

jborean93 replied:
Up to you, I don't really want to take away from all the work you've done so far. If we were to make it into a separate module we can close this one. If we wanted to include the functionality into this module we can do 2 things

1. Keep moving forward with this and at a future point in time implement the PInvoke stuff in a separate PR
2. Just do the PInvoke stuff in 1 PR

Honestly I will leave it up to you, you've done a lot of work here and I don't take away from what's currently here.

rnsc replied:
Ok, I'll think about it over the week-end!
In any case, I needed these changes internally in my company, so the work is not lost and we use a local version in our repo.

jborean93 reviewed (changes requested):
Nearly there, there are just 3 things I see need to be updated:

1. The docs should document the sub options for `recovery_actions` so people can know what actions are available and the options for those actions
2. The return value for `recovery` should be in the same format (as close as possible) to the input `recovery_actions` list. This keeps things aligned between the 2
3. We should handle a list that is less than 3 by just padding Take No Action up to 3, i.e. an empty list means Take No Action for all 3, a list with 1 entry means the remaining 2 are Take No Action and so forth.

Apart from that everything else looks great and we should be able to merge it in soon.

jborean93 commented:
While I do have a script to help migrate these changes I am currently working on refactoring the `win_service` module to include the features added in this PR. If you still wish to get this in some way please let me know and I'll run the migration tool on this. The end location for this module is in https://github.com/ansible-collections/ansible.windows.

jborean93 commented:
It too me longer than I was hoping but here is the PR that adds recovery actions to win_service https://github.com/ansible-collections/ansible.windows/pull/40

rnsc commented:
> It too me longer than I was hoping but here is the PR that adds recovery actions to win_service [ansible-collections/ansible.windows#40](https://github.com/ansible-collections/ansible.windows/pull/40)

Thank you @jborean93 !
